### ISSUE TITLE
Special treatment of DEFAULT prohibits parsing of some valid definitions

### ISSUE THREAD

**Issue by mvangeest** [JakeWheat/hssqlppp#16]
Currently, the default value of a table attribute is treated specially (compared to other column constraints). It is [stored in the `def` field of `AttributeDef`](https://github.com/JakeWheat/hssqlppp/blob/6359814e6741ff7aba0edd79b2e935aa21735e03/hssqlppp/src/Database/HsSqlPpp/Internals/AstInternal.ag#L1068) and [parsed separately](https://github.com/JakeWheat/hssqlppp/blob/6359814e6741ff7aba0edd79b2e935aa21735e03/hssqlppp/src/Database/HsSqlPpp/Internals/ParseInternal.lhs#L670). I understand that this design facilitates type checking, but it prevents the processing of several valid attribute definitions.

According to the [grammar for CREATE TABLE in the PostgreSQL documentation](https://www.postgresql.org/docs/9.6/static/sql-createtable.html), the `DEFAULT` keyword introduces a constraint like any other. This means we can place it after other constraints:

```sql
CREATE TABLE t (a integer NOT NULL DEFAULT 1);
```

We can also name it explicitly:

```sql
CREATE TABLE t (a integer CONSTRAINT adefault DEFAULT 1);
```

Both these definitions are accepted by PostgreSQL (and the former is accepted by many other database systems), but rejected by the hssqlppp parser.

I would like to use hssqlppp to parse some existing table definitions, which unfortunately contain constraints in many different orders. For my application, it is important that pretty-printing a parsed definition does not significantly alter the SQL. I therefore recommend treating `DEFAULT` just like the other column constraints (i.e. adding a `DefaultConstraint` constructor to `RowConstraint`, parsing it inside the `choice`, etc.). This allows me to output something very similar to what I've read, and I think it might be reasonably easy to implement. It does mean the default value will have to be retrieved using `listToMaybe [v | DefaultConstraint _ _ v <- constraints]`.

**Comment by JakeWheat** [comment]
It makes no sense to me whatsoever that default is a constraint, but that is the sql standard and how most sql products work. I don't think the change you suggest should affect the difficulty of typechecking much. Can you make a patch?

**Comment by mvangeest** [comment]
I'll take a look. I've been lucky to work with UUAGC during a course here at Utrecht University (UU) in 2015 - and I really enjoyed the concept, but the tooling support wasn't the greatest. It looks like Cabal still doesn't have native UUAGC preprocessing support (uuagc-cabal is only a workaround), but I see you've got a Makefile to do the hard work. I'll see if I can create a patch in the next few days (possibly weeks).

**Comment by mvangeest** [comment]
Fixed by #17. Thanks!
